### PR TITLE
Add Flax Dropout guide

### PR DESCRIPTION
**Pull request by 8bitmp3**
- Add a dedicated _Dropout_ guide.
- Covers PRNG key usage, model creation (with `nn.Module` and `nn.Dropout`), initialization, the forward pass, and the training step (with `TrainStep`).
- Includes examples from `flax/examples` like a Transformer-based model trained on the WMT dataset (with dropout and attention dropout).

~~Shorten the [Flax - The Sharp Bits](https://flax.readthedocs.io/en/latest/notebooks/flax_sharp_bits.html) by migrating some of the dropout instructions to the new guide.~~ (To be done in a separate PR)

Live preview: https://flax--2675.org.readthedocs.build/en/2675/guides/dropout.html

### CODE REVIEW

**Comment by 8bitmp3** [comment]
Thanks @marcvanzee 👍 

@cgarciae Your feedback would be super welcome 👍

File: docs/guides/dropout.rst [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+To create a model, subclass :meth:`flax.linen.Module`, and then use
+:meth:`flax.linen.Dropout` to add a dropout layer. Recall that
+:meth:`flax.linen.Module` is the
+`base class for all neural network modules <https://flax.readthedocs.io/en/latest/api_reference/flax.linen.html#module>`__
+, and all layers and models are subclassed from it.
+
+In :meth:`flax.linen.Dropout`, the ``deterministic`` argument is ``None`` by
+default.
+
+* If it's set to ``False``, the inputs are scaled by ``1 / (1 - dropout_rate)`` and masked.
```
Comment (marcvanzee) [inline docs/guides/dropout.rst:55]:
```suggestion
* If it's set to ``False``, the inputs are masked (i.e. set to zero) with a probablity set by ``rate``, and the remaining inputs are scaled by ``1 / (1 - rate)``, which to ensure the means of the inputs are preserved.
```

File: docs/guides/dropout.rst [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+* When it's ``True``, no mask is applied (the dropout is turned off).
```
Comment (marcvanzee) [inline docs/guides/dropout.rst:56]:
```suggestion
* When it's ``True``, no mask is applied (the dropout is turned off), and the inputs are returned as-is.
```

File: docs/guides/dropout.rst [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+
+Flax provides an implicit way of handling PRNG key streams via Flax
+:meth:`flax.linen.Module`'s :meth:`flax.linen.Module.make_rng` helper function.
+It allows the code in Flax ``Module``s (or their sub-``Module``s) to
+"pull PRNG keys". ``make_rng`` guarantees to provide a unique key each time you
+call it.
+
```
Comment (marcvanzee) [inline docs/guides/dropout.rst:63]: I suggest removing this paragraph since it contains the same content as the next paragraph, and I think the next one is a bit more clear.

```suggestion
```

File: docs/guides/dropout.rst [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+**Note:** Internally, :meth:`flax.linen.Dropout` makes use of :meth:`flax.linen.Module.make_rng`
+to create a key for dropout. You can check out the
+`source code <https://github.com/google/flax/blob/5714e57a0dc8146eb58a7a06ed768ed3a17672f9/flax/linen/stochastic.py#L72>`__.
+
+* Every time :meth:`Module.make_rng` is called (which is done implicitly in
```
Comment (marcvanzee) [inline docs/guides/dropout.rst:68]: Dropout does not call this implicitly, it calls it directly.

```suggestion
* Every time :meth:`Module.make_rng` is called (which is done in
```

File: docs/guides/dropout.rst [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+  :meth:`flax.linen.Dropout`), you get a new PRNG key split from the main/root pseudorandom
+  number generator (PRNG) key.
```
Comment (marcvanzee) [inline docs/guides/dropout.rst:70]: Maybe write "pseudorandom number generator (PRNG, or simply RNG)" once at the beginning of the tutorial, and then use RNG throughout the tutorial?


```suggestion
  :meth:`flax.linen.Dropout`), you get a new PRNG key split from the RNG key that is passed at the top level (see [next section](#add-link) for more details).
```

File: docs/guides/dropout.rst [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+      # When the `deterministic` flag is `True`, dropout is turned off. #!
+      x = nn.Dropout(rate=0.5, deterministic=not self.training)(x) #!
+      return x
+
+After adding ``Dropout`` to your model, instantiate the model with
+``training=True``.
+
+For example, when creating a ``TrainState``, the code will include::
+
+  my_model = MyModel(num_neurons=3, training=True)
```
Comment (marcvanzee) [inline docs/guides/dropout.rst:106]: I wouldn't mention TrainState here since TrainState does not store a module, only the apply_fn.

```suggestion
``training=True`` in order to apply Dropout::

  my_model = MyModel(num_neurons=3, training=True)
```

File: docs/guides/dropout.rst [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+Initializing the model
+**********************
+
+The PRNG key provided to :meth:`flax.linen.init()` is the one that seeds the
+``'params'`` PRNG key stream.
+
+Initialize the model with the ``params'`` key::
+
+  x = jax.random.uniform(key=main_key, shape=(3, 4, 4))
+  variables = my_model.init(params_key, x)
```
Comment (marcvanzee) [inline docs/guides/dropout.rst:117]: This is not correct. If you construct `my_model` as you did in the previous section, then you should pass an RNG for `dropout` here. I would also explain that you don't need to pass it if Dropout is disabled.

Maybe you can use a side-by-side diff view here as well?

```suggestion
When applying Dropout, you should pass the RNG key for it to :meth:`flax.linen.init()` with name `dropout`. If do you not apply Dropout, you don't have to pass it. Since you are then only passing the RNG for initializing parameters, you can abbreviate ``{'params': params_rng}`` with simply ``params_rng`` (see :meth:`flax.linen.init()` for more details).

.. codediff::
  :title_left: Dropout disabled
  :title_right: Dropout enabled
  :sync:
  
  rng = random.PRNGKey(0)
  data_rng, params_rng, rng = random.split(rng, 3)
  x = jax.random.uniform(data_rng, shape=(3, 4, 4))
  
  my_model = MyModel(num_neurons=3, training=False)
  variables = model_nodropout.init(params_rng, x)

  ---
  rng = random.PRNGKey(0)
  data_rng, params_rng, dropout_rng, rng = random.split(rng, 4)
  x = jax.random.uniform(data_rng, shape=(3, 4, 4))
  init_rngs = {'params': params_key, 'dropout': dropout_key}
  my_model = MyModel(num_neurons=3, training=True)
  variables = model_dropout.init(init_rngs, x)

```

File: docs/guides/dropout.rst [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+
+.. codediff::
+  :title_left: No Dropout
+  :title_right: With Dropout
+  :sync:
+
+  y = my_model.apply(variables, x)
+
+  ---
+  y = my_model.apply(variables, x, rngs={'dropout': dropout_key}) #!
```
Comment (marcvanzee) [inline docs/guides/dropout.rst:135]: I would clarify here that you should construct the Module here with the right keyword arguments.

```suggestion
  my_model = MyModel(num_neurons=3, training=False)
  y = my_model.apply(variables, x)

  ---
  my_model = MyModel(num_neurons=3, training=True)
  y = my_model.apply(variables, x, rngs={'dropout': dropout_key}) #!
```

File: docs/guides/dropout.rst [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+
+During evaluation, you don't need to pass the ``'dropout'`` key to ``rngs`` in
+:method:`flax.linen.apply()`.
```
Comment (marcvanzee) [inline docs/guides/dropout.rst:138]:
```suggestion
Typically, the left-hand side is used during training, and the right-hand side during evaluation.
```

File: docs/guides/dropout.rst [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+
```
Comment (marcvanzee) [inline docs/guides/dropout.rst:139]: I would also explain here that even if you apply Dropout somewhere in a submodule, you still only have to pass the RNG at the top. This is something that users are often confused about.

**Review by marcvanzee — CHANGES_REQUESTED** [review]
Great work! I think this will be an awesome tutorial, just needs a bit of work to get it really perfect.

File: docs/guides/dropout.rst [hunk 11]
```diff
@@ -0,0 +1,147 @@
...
+* :meth:`Module.make_rng` guarantees full reproducibility.
+
+.. codediff::
+  :title_left: No Dropout
+  :title_right: With Dropout
+  :sync:
+
+  class MyModel(nn.Module):
+    num_neurons: int
+    training: bool
```
Comment (cgarciae) [inline docs/guides/dropout.rst:80]:
```suggestion

```

File: docs/guides/dropout.rst [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+  ---
+  class MyModel(nn.Module):
+    num_neurons: int
+    training: bool
+    @nn.compact
+    def __call__(self, x):
+      x = nn.Dense(self.num_neurons)(x)
+      # Set the dropout layer with a rate of 50%. #!
+      # When the `deterministic` flag is `True`, dropout is turned off. #!
+      x = nn.Dropout(rate=0.5, deterministic=not self.training)(x) #!
```
Comment (cgarciae) [inline docs/guides/dropout.rst:98]: Ideally we keep passing training state via a call argument as the user can control it via `Module.apply` / `TrainState.apply_fn`, else you'd be forced to either use mutable state (which has pitfalls) or recreating entire `Module` / `TrainState`.
```suggestion

    @nn.compact
    def __call__(self, x, train: bool):
      x = nn.Dense(self.num_neurons)(x)
      # Set the dropout layer with a rate of 50%. #!
      # When the `deterministic` flag is `True`, dropout is turned off. #!
      x = nn.Dropout(rate=0.5, deterministic=not train)(x) #!
```

Comment (8bitmp3) [reply docs/guides/dropout.rst:98]: @cgarciae Can we remind users that they can use both `train` and `training`? This may start getting confusing especially for less experienced users. cc @levskaya @marcvanzee (API design)

File: docs/notebooks/flax_sharp_bits.md [hunk 13]
```diff
@@ -28,76 +28,15 @@ First, install and/or update Flax:
 
 ### TL;DR
 
-When working on a model with dropout (subclassed from [Flax `Module`](https://flax.readthedocs.io/en/latest/guides/flax_basics.html#module-basics)), add the `'dropout'` PRNGkey only during the forward pass.
+- When initializing the model with [`flax.linen.init()`](https://flax.readthedocs.io/en/latest/api_reference/flax.linen.html#init-apply), there's no need to pass in an extra `'dropout'` pseudorandom number generator (PRNG) key—just the `'params'` key like in a "simpler" model.
```
[the remark this answers is not in the document]
Comment (8bitmp3) [reply docs/notebooks/flax_sharp_bits.md:31]: Excellent @marcvanzee .

- Rephrased/refactored some of your suggestions.
- Added small Python snippets to aid users who prefer examples ("show me, don't tell me").
- Step 1: Set `deterministic=False` to apply `Dropout` -> Step 2: In `flax.linen.Module.apply`, provide `rngs={'dropout':...`.
  
  Note that all this stuff does kind of repeat what it says in the new Dropout Guide, but new users can benefit from repetition. This "sequence" (do this in Step 1, then this in Step 2 may not be very intuitive).

- Added **Training:** and **Inference:** to two sections. for those users who are searching for keywords when scanning the document. This should improve UX.

Below is the revamped chapter on "Dropout and randomness" in Flax Sharp Bits (all the way until the end):

---

## 🔪 `flax.linen.Dropout` layer and randomness


- When initializing the model with [`flax.linen.init()`](https://flax.readthedocs.io/en/latest/api_reference/flax.linen.html#init-apply), you don't need to pass in an extra `'dropout'` pseudorandom number generator (PRNG) key—just the `'params'` key like in a "regular" model.

  ```python
  variables = my_model.init(your_params_key, x)
  ```

- **Training**: When you need to have the [`Dropout`](https://flax.readthedocs.io/en/latest/api_reference/_autosummary/flax.linen.Dropout.html) layer applied to your model, you set the `deterministic` parameter to `False` (usually for training). When calling [`flax.linen.apply()`](https://flax.readthedocs.io/en/latest/api_reference/flax.linen.html#init-apply) for the the forward pass, you need to provide a keyword argument `rngs={'dropout': dropout_key}` to the top-level [`Module`](https://flax.readthedocs.io/en/latest/guides/flax_basics.html#module-basics). That is, in [`flax.linen.Dropout`](https://flax.readthedocs.io/en/latest/api_reference/_autosummary/flax.linen.Dropout.html), 

  ```python
  y = my_model.apply(variables, x, rngs={'dropout': dropout_key})
  ```

- **Inference:** When you don't need to have the [`Dropout`](https://flax.readthedocs.io/en/latest/api_reference/_autosummary/flax.linen.Dropout.html) layer applied–for example, during inference–you don't have to pass the `rngs` keyword argument to [`flax.linen.apply()`](https://flax.readthedocs.io/en/latest/api_reference/flax.linen.html#init-apply).

- You only have to pass the PRNG key to the top-level [`Module`](https://flax.readthedocs.io/en/latest/guides/flax_basics.html#module-basics), even if you define the [`Dropout`](https://flax.readthedocs.io/en/latest/api_reference/_autosummary/flax.linen.Dropout.html) layer in a nested sub-[`Module`](https://flax.readthedocs.io/en/latest/guides/flax_basics.html#module-basics). **Flax passes down PRNG keys to sub-`Module`s automatically.**

### PRNGs in JAX and Flax

The [dropout](https://jmlr.org/papers/volume15/srivastava14a/srivastava14a.pdf) stochastic regularization technique is a random operation, requiring a PRNG state, and Flax (like JAX) uses [Threefry](https://github.com/google/jax/blob/main/docs/jep/263-prng.md) PRNG that is splittable.

PRNGs in JAX are covered in [🔪 JAX - The Sharp Bits 🔪 Randomness and PRNG keys](https://jax.readthedocs.io/en/latest/notebooks/Common_Gotchas_in_JAX.html#random-numbers).

PRNG key streams are handled implicitly via [`Module`](https://flax.readthedocs.io/en/latest/guides/flax_basics.html#module-basics)'s [`flax.linen.Module.make_rng`](https://flax.readthedocs.io/en/latest/api_reference/flax.linen.html#flax.linen.Module.make_rng), which enables the code in `Module`s or sub-`Module`s to "pull PRNG keys". `make_rng` guarantees to provide a unique key each time you call it.

### Example

Refer to the **Flax guide on [applying dropout](https://flax.readthedocs.io/en/latest/guides/dropout.html)** for a detailed explanation and an example.

File: docs/guides/dropout.rst [hunk 14]
```diff
@@ -0,0 +1,147 @@
+Applying dropout
+================
+
+This guide provides an overview of how to apply
+`dropout <https://jmlr.org/papers/volume15/srivastava14a/srivastava14a.pdf>`__
+using :meth:`flax.linen.Dropout`
```
[the remark this answers is not in the document]
Comment (8bitmp3) [reply docs/guides/dropout.rst:6]: Thanks for the catch, added.

File: docs/guides/dropout.rst [hunk 15]
```diff
@@ -0,0 +1,147 @@
...
+
+Split the PRNG
+**************
+
+Dropout is a random operation, requiring a PRNG state. Flax (like JAX) uses
+`Threefry <https://github.com/google/jax/blob/main/docs/jep/263-prng.md>`__ PRNG
+that is splittable.
+
+**Note:** Recall that JAX has an explicit way of giving you PRNG keys:
+you can fork the main PRNG state (such as ``key = jax.random.PRNGKey(seed=0)``)
```
[the remark this answers is not in the document]
Comment (8bitmp3) [reply docs/guides/dropout.rst:29]: I agree that it's not common to have the argument name, but just in case some users wonder what `0` stands for (it could be any number), we should keep this IMO.

File: docs/guides/dropout.rst [hunk 16]
```diff
@@ -0,0 +1,147 @@
...
+  import jax.numpy as jnp
+  import jax
+  import optax
+
+Split the PRNG
+**************
+
+Dropout is a random operation, requiring a PRNG state. Flax (like JAX) uses
+`Threefry <https://github.com/google/jax/blob/main/docs/jep/263-prng.md>`__ PRNG
+that is splittable.
```
[the remark this answers is not in the document]
Comment (8bitmp3) [reply docs/guides/dropout.rst:26]:
```
Since dropout is a random operation, it requires a pseudorandom number generator
(PRNG) state. Flax uses JAX's (splittable) PRNG keys, which have a number of desirable
properties for neutral networks. To learn more, refer to the
`Pseudorandom numbers in JAX tutorial <https://jax.readthedocs.io/en/latest/jax-101/05-random-numbers.html`__.
```

LMKWYT

("Pseudorandom" is one word, I think - we should fix this on JAX docs)

File: docs/guides/dropout.rst [hunk 17]
```diff
@@ -0,0 +1,147 @@
...
+
+**Note:** Recall that JAX has an explicit way of giving you PRNG keys:
+you can fork the main PRNG state (such as ``key = jax.random.PRNGKey(seed=0)``)
+into multiple new PRNG keys with ``key, subkey = jax.random.split(key)``. You
+can refresh your memory in
+`🔪 JAX - The Sharp Bits 🔪 Randomness and PRNG keys <https:/t/jax.readthedocs.io/en/latest/notebooks/Common_Gotchas_in_JAX.html#random-numbers>`__.
```
[the remark this answers is not in the document]
Comment (8bitmp3) [reply docs/guides/dropout.rst:32]: I added your suggested link to the JAX 101 tutorial earlier, so let's keep both links. The JAX Sharp Bits is a treasure some users may not be aware of.

Fixed the link 👍

File: docs/guides/dropout.rst [hunk 18]
```diff
@@ -0,0 +1,147 @@
...
+
+You split the PRNG into keys, including one for dropout, as follows::
+
+  seed = 0
+  root_key = jax.random.PRNGKey(seed=seed)
+  main_key, params_key, dropout_key = jax.random.split(key=root_key, num=3)
+
+Each of the Flax PRNG streams has a name, such as the ``'params'`` stream for
+initializing parameters, and ``'dropout'`` for dropout.
```
[the remark this answers is not in the document]
Comment (8bitmp3) [reply docs/guides/dropout.rst:41]: Good idea.

With some minor refactoring of your suggestion:

```
In Flax, you provide PRNG streams with names, so that you can use them later in
your ``Module``. For example, you pass the stream ``'params'`` for
initializing parameters, and ``'dropout'`` for applying :meth:`flax.linen.Dropout`.
```

File: docs/guides/dropout.rst [hunk 19]
```diff
@@ -0,0 +1,147 @@
...
+
+Defining a model with ``Dropout``
+*********************************
+
+To create a model, subclass :meth:`flax.linen.Module`, and then use
+:meth:`flax.linen.Dropout` to add a dropout layer. Recall that
+:meth:`flax.linen.Module` is the
+`base class for all neural network modules <https://flax.readthedocs.io/en/latest/api_reference/flax.linen.html#module>`__
+, and all layers and models are subclassed from it.
```
[the remark this answers is not in the document]
Comment (8bitmp3) [reply docs/guides/dropout.rst:50]: Just trying to keep each line in this reStructuredText file not too long, and the link is just too long, so it looks like:

```
To create a model, subclass :meth:`flax.linen.Module`, and then use
:meth:`flax.linen.Dropout` to add a dropout layer. Recall that
:meth:`flax.linen.Module` is the
`base class for all neural network modules <https://flax.readthedocs.io/en/latest/api_reference/flax.linen.html#module>`__,
and all layers and models are subclassed from it.
```

I am moving the last comma though 😄 Thanks

File: docs/guides/dropout.rst (hunk 1 above)
Comment (8bitmp3) [reply docs/guides/dropout.rst:55]: Awesome. Thanks!

Applied with some minor changes:

```

* If it's set to ``False``, the inputs are masked (that is, set to zero) with able
  probability set by ``rate``. And the remaining inputs are scaled by ``1 / (1 - rate)``,
  which ensures that the means of the inputs are preserved.
```

cc @levskaya @marcvanzee The `None` is a bit of an odd default bool for `deterministic` 🤔

File: docs/guides/dropout.rst (hunk 2 above)
Comment (8bitmp3) [reply docs/guides/dropout.rst:56]: 👍  Thanks

File: docs/guides/dropout.rst (hunk 3 above)
Comment (8bitmp3) [reply docs/guides/dropout.rst:63]: I think some users may benefit from some repetition. Maybe merging the two paragraphs under one **Note:** may shorten this section:

```
**Note:** Flax provides an implicit way of handling PRNG key streams via Flax
:meth:`flax.linen.Module`'s :meth:`flax.linen.Module.make_rng` helper function.
It allows the code in Flax ``Module``s (or their sub-``Module``s) to
"pull PRNG keys". ``make_rng`` guarantees to provide a unique key each time you
call it. Internally, :meth:`flax.linen.Dropout` makes use of :meth:`flax.linen.Module.make_rng`
to create a key for dropout. You can check out the
`source code <https://github.com/google/flax/blob/5714e57a0dc8146eb58a7a06ed768ed3a17672f9/flax/linen/stochastic.py#L72>`__.
```

The internals and `make_rng` is another new guide (idea).

File: docs/guides/dropout.rst (hunk 4 above)
Comment (8bitmp3) [reply docs/guides/dropout.rst:68]: Thanks! Done.

File: docs/guides/dropout.rst (hunk 3 above)
Comment (8bitmp3) [reply docs/guides/dropout.rst:63]: Also, I think @andsteing pointed me to the source code, which is awesome. Some Flax users may find this useful too.

File: docs/guides/dropout.rst (hunk 5 above)
Comment (8bitmp3) [reply docs/guides/dropout.rst:70]: Thanks for the catch. Not 100% sure how to do section links, so this should do for now:

```
* Every time :meth:`Module.make_rng` is called (which is done in
  :meth:`flax.linen.Dropout`), you get a new PRNG key split from the main/root PRNG key.
```

File: docs/guides/dropout.rst (hunk 6 above)
Comment (8bitmp3) [reply docs/guides/dropout.rst:106]: Thanks! Done.

```
After adding ``Dropout`` to your model, instantiate the model with
``training=True`` to apply dropout::

  my_model = MyModel(num_neurons=3, training=True)
```

File: docs/guides/dropout.rst [hunk 20]
```diff
@@ -0,0 +1,147 @@
...
+
+In :meth:`flax.linen.Dropout`, the ``deterministic`` argument is ``None`` by
+default.
```
[the remark this answers is not in the document]
Comment (8bitmp3) [reply docs/guides/dropout.rst:53]: Excellent! 

```
In :meth:`flax.linen.Dropout`, the ``deterministic`` argument is required to be passed as a
keyword argument, either when constructing the :meth:`flax.linen.Module` or when calling
:meth:`flax.linen.init()`/:meth:`flax.linen.apply()` on a constructed ``Module``.
(Refer to :meth:`flax.linen.module.merge_param` for more details.)
```

File: docs/guides/dropout.rst (hunk 7 above)
Comment (8bitmp3) [reply docs/guides/dropout.rst:117]: Sounds good!

With some minor changes:

```
- If you enable :meth:`flax.linen.Dropout` in your model, you need to pass the
  ``'dropout'`` PRNG key together with the ``'params'`` key.
- If you have ``Dropout`` disabled, the ``'params'`` PRNG key is enough for
  :meth:`flax.linen.init()`.

Refer to the :meth:`flax.linen.init()` API docs for more details.

.. codediff::
  :title_left: Dropout disabled
  :title_right: Dropout enabled
  :sync:

  my_model_without_dropout = MyModel(num_neurons=3, training=False)

  x = jax.random.uniform(key=main_key, shape=(3, 4, 4))
  variables = my_model.init(params_rngs, x)
  ---
  my_model_with_dropout = MyModel(num_neurons=3, training=True) #!
  init_rngs = {'params': params_key, 'dropout': dropout_key} #!
  x = jax.random.uniform(key=main_key, shape=(3, 4, 4))
  variables = my_model.init(init_rngs, x) #!
```

File: docs/guides/dropout.rst (hunk 8 above)
Comment (8bitmp3) [reply docs/guides/dropout.rst:135]: Good idea.

```
.. codediff::
  :title_left: Dropout disabled
  :title_right: Dropout enabled
  :sync:

  my_model_without_dropout = MyModel(num_neurons=3, training=False)
  y = my_model.apply(variables, x)

  ---
  my_model_with_dropout = MyModel(num_neurons=3, training=True) #!
  y = my_model.apply(variables, x, rngs={'dropout': dropout_key}) #!
```

File: docs/guides/dropout.rst (hunk 11 above)
Comment (8bitmp3) [reply docs/guides/dropout.rst:80]: Good catch! Thanks @cgarciae

File: docs/guides/dropout.rst (hunk 12 above)
Comment (8bitmp3) [reply docs/guides/dropout.rst:98]: > Ideally we keep passing training state via a call argument as the user can control it via Module.apply / TrainState.apply_fn, else you'd be forced to either use mutable state (which has pitfalls) or recreating entire Module / TrainState.

Done. Thanks @cgarciae

**Comment by 8bitmp3** [comment]
@marcvanzee @cgarciae Thanks. PTAL 👍

**Review by cgarciae — COMMENTED** [review]
PTAL

**Comment by 8bitmp3** [comment]
Thanks for the help @cgarciae :+1: 

@marcvanzee some of @cgarciae's review comments got stuck in "pending", they are out now (see above). We have gone over them and the new commits are on the way. Some changes will involve some changes to your code suggestions.

File: docs/guides/dropout.rst [hunk 21]
```diff
@@ -0,0 +1,165 @@
...
+
+  my_model = MyModel(num_neurons=3, training=False)
+
+  x = jax.random.uniform(key=main_key, shape=(3, 4, 4))
+  variables = my_model.init(params_rngs, x)
+  ---
+  my_model = MyModel(num_neurons=3, training=True) #!
+  init_rngs = {'params': params_key, 'dropout': dropout_key} #!
+  x = jax.random.uniform(key=main_key, shape=(3, 4, 4))
+  variables = my_model.init(init_rngs, x) #!
```
[the remark this answers is not in the document]
Comment (8bitmp3) [reply docs/guides/dropout.rst:133]: @cgarciae Thanks again. Can we replace `main_key` -> `params_key`?

```
  root_key = jax.random.PRNGKey(seed=seed)
  main_key, params_key, dropout_key = jax.random.split(key=root_key, num=3)
  ...
  my_model = MyModel(num_neurons=3)
  x = jnp.empty((3, 4, 4))
  variables = my_model.init(params_key, x, training=False) #!
  params = variables['params']
```

**Comment by 8bitmp3** [comment]
@marcvanzee fyi working on adding the training step

File: docs/notebooks/flax_sharp_bits.md (hunk 13 above)
[the remark this answers is not in the document]
Comment (8bitmp3) [reply docs/notebooks/flax_sharp_bits.md]: @marcvanzee The changes to the Sharp Bits will be implemented in a separate PR.

File: docs/guides/dropout.rst (hunk 7 above)
Comment (8bitmp3) [reply docs/guides/dropout.rst]: Revamped `init()`, etc, no longer using `  init_rngs = {'params': params_key, 'dropout': dropout_key}`.

File: docs/guides/dropout.rst (hunk 9 above)
Comment (8bitmp3) [reply docs/guides/dropout.rst]: Only using the evaluation comment. Added a separate training step section.

File: docs/guides/dropout.rst [hunk 22]
```diff
@@ -0,0 +1,165 @@
...
+  :title_left: Dropout disabled
+  :title_right: Dropout enabled
+  :sync:
+
+  my_model = MyModel(num_neurons=3, training=False)
+  y = my_model.apply(variables, x)
+
+  ---
+  my_model = MyModel(num_neurons=3, training=True) #!
+  y = my_model.apply(variables, x, rngs={'dropout': dropout_key}) #!
```
[the remark this answers is not in the document]
Comment (8bitmp3) [reply docs/guides/dropout.rst]: :+1:  Thanks again!

File: docs/guides/dropout.rst [hunk 23]
```diff
@@ -0,0 +1,239 @@
...
+  :title_left: No Dropout
+  :title_right: With Dropout
+  :sync:
+
+
+  y = my_model.apply({'params': params}, x)
+
+   ---
+  # Dropout is enabled with `training=True` (that is, `deterministic=False`). #!
+  y = my_model.apply({'params': params}, x, training=True, rngs={'dropout': dropout_key}) #!
```
Comment (8bitmp3) [inline docs/guides/dropout.rst:164]:
```suggestion
  y = my_model.apply({'params': params}, x)

  ---
  # Dropout is enabled with `training=True` (that is, `deterministic=False`). #!
  y = my_model.apply({'params': params}, x, training=True, rngs={'dropout': dropout_key}) #!
```

File: docs/guides/dropout.rst [hunk 24]
```diff
@@ -0,0 +1,239 @@
...
+
+
+  y = my_model.apply({'params': params}, x)
+
+  ---
+  # Dropout is enabled with `training=True` (that is, `deterministic=False`). #!
+  y = my_model.apply({'params': params}, x, training=True, rngs={'dropout': dropout_key}) #!
+
+During evaluation, use the above code with no dropout enabled (this means you do
+not have to pass a RNG either).
```
Comment (8bitmp3) [inline docs/guides/dropout.rst:167]:
```suggestion
not have to pass a PRNG key either).
```

**Comment by 8bitmp3** [comment]
Reviewed the code and text with @marcvanzee. Committed and pushed the changes to the branch. cc @cgarciae. Amendments to the Sharp Bits will be in a separate PR.

File: docs/guides/dropout.rst [hunk 25]
```diff
@@ -0,0 +1,240 @@
...
+
+.. codediff::
+  :title_left: No Dropout
+  :title_right: With Dropout
+  :sync:
+
+
+  y = my_model.apply({'params': params}, x)
+
+   ---
```
Comment (marcvanzee) [inline docs/guides/dropout.rst:160]:
```suggestion
  ---
```

**Comment by marcvanzee** [comment]
LGTM -- thanks a lot for this @8bitmp3, looks really awesome!! 🥳 

(Please fix the doc build failures)

**Comment by 8bitmp3** [comment]
All checks :heavy_check_mark: @cgarciae PTAL. Thank you.

Live preview: https://flax--2675.org.readthedocs.build/en/2675/guides/dropout.html (build successful @marcvanzee).

File: docs/guides/dropout.rst [hunk 26]
```diff
@@ -0,0 +1,299 @@
...
+
+More Flax examples that use randomness
+**************************************
+
+* Applying word dropout to a batch of input IDs in a
+  `text classification <https://github.com/google/flax/blob/main/examples/sst2/models.py>`__
+  context.
+
+* Defining a prediction token in a decoder of a
+  `sequence-to-sequence model <https://github.com/google/flax/blob/main/examples/seq2seq/models.py>`__.
```
Comment (marcvanzee) [inline docs/guides/dropout.rst:299]: Small suggestion: "using randomness" is a bit odd since any ML model uses randomness when initializing the parameters. Also the sst model actually uses dropout, but a custom one, so i would put it under the dropout section.

```suggestion
Flax examples with dropout
**************************

* A `Transformer-based model <https://github.com/google/flax/blob/main/examples/wmt/models.py>`__
  trained on the WMT Machine Translation dataset. This example uses dropout and attention dropout.
  
* Applying word dropout to a batch of input IDs in a
  `text classification <https://github.com/google/flax/blob/main/examples/sst2/models.py>`__
  context. This example defines a custom Dropout layer.

More Flax examples that use `make_rng()`
**************************************

* Defining a prediction token in a decoder of a
  `sequence-to-sequence model <https://github.com/google/flax/blob/main/examples/seq2seq/models.py>`__.
```

Comment (8bitmp3) [reply docs/guides/dropout.rst]: Done:

```
Flax examples with dropout
**************************

* A `Transformer-based model <https://github.com/google/flax/blob/main/examples/wmt/models.py>`__
  trained on the WMT Machine Translation dataset. This example uses dropout and attention dropout.

* Applying word dropout to a batch of input IDs in a
  `text classification <https://github.com/google/flax/blob/main/examples/sst2/models.py>`__
  context. This example uses a custom :meth:`flax.linen.Dropout` layer.

More Flax examples that use Module ``make_rng()``
*************************************************

* Defining a prediction token in a decoder of a
  `sequence-to-sequence model <https://github.com/google/flax/blob/main/examples/seq2seq/models.py>`__.

```

Thanks @marcvanzee 

Also changed the previous section title from "Training" to "`TrainState` and the training step".

**Comment by 8bitmp3** [comment]
@cgarciae PTAL.

**Review by cgarciae — APPROVED** [review]
Thanks @8bitmp3! Looks great, lets merge this!
